### PR TITLE
[mlir][openmp] - Fix crash in OpenMPIRBuilder when converting to LLVMIR

### LINKED ISSUE #84606: [flang][openmp] - flang-new crashes when omp.task and omp.target are used in the same function
Command
```
flang-new   -O2    -fopenmp --offload-arch=gfx90a  -D__OFFLOAD_ARCH_gfx90a__ f.f95 -o f.out
```
Crash signature
```
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /work/prbhanda/git/bhandarkar-pranav/build/install_18.0-0-25_Jan_24_23_23/bin/flang-new -fc1 -triple amdgcn\
-amd-amdhsa -emit-llvm-bc -D __OFFLOAD_ARCH_gfx90a__ -fopenmp -mrelocation-model pic -pic-level 2 -target-cpu gfx90a -fopenmp-host-ir-\
file-path /tmp/f-6566c4.bc -fopenmp-is-target-device -mframe-pointer=none -O2 -o /tmp/f-gfx90a-e38eba.bc -x f95-cpp-input f.f95
 #0 0x00007f98c95dcac0 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/work/prbhanda/git/bhandarkar-pranav/build/install_18.0-0-\
25_Jan_24_23_23/bin/../lib/libLLVMSupport.so.19.0git+0x1dcac0)
 #1 0x00007f98c95d9ecf llvm::sys::RunSignalHandlers() (/work/prbhanda/git/bhandarkar-pranav/build/install_18.0-0-25_Jan_24_23_23/bin/.\
./lib/libLLVMSupport.so.19.0git+0x1d9ecf)
 #2 0x00007f98c95da025 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007f98c8c42520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007f98c70cee2c llvm::CodeExtractorAnalysisCache::CodeExtractorAnalysisCache(llvm::Function&) (/work/prbhanda/git/bhandarkar-pr\
anav/build/install_18.0-0-25_Jan_24_23_23/bin/../lib/../lib/libLLVMTransformUtils.so.19.0git+0xcee2c)
 #5 0x00007f98c2d4cbc5 llvm::OpenMPIRBuilder::finalize(llvm::Function*) (/work/prbhanda/git/bhandarkar-pranav/build/install_18.0-0-25_\
Jan_24_23_23/bin/../lib/../lib/libLLVMFrontendOpenMP.so.19.0git+0x6ebc5)
 #6 0x00007f98ca3b2c31 mlir::LLVM::ModuleTranslation::~ModuleTranslation() (/work/prbhanda/git/bhandarkar-pranav/build/install_18.0-0-\
25_Jan_24_23_23/bin/../lib/../lib/libMLIRTargetLLVMIRExport.so.19.0git+0x24c31)
 #7 0x00007f98ca3c9c13 mlir::translateModuleToLLVMIR(mlir::Operation*, llvm::LLVMContext&, llvm::StringRef) (/work/prbhanda/git/bhanda\
rkar-pranav/build/install_18.0-0-25_Jan_24_23_23/bin/../lib/../lib/libMLIRTargetLLVMIRExport.so.19.0git+0x3bc13)
 #8 0x00007f98cbdc8fef Fortran::frontend::CodeGenAction::generateLLVMIR() (/work/prbhanda/git/bhandarkar-pranav/build/install_18.0-0-2\
5_Jan_24_23_23/bin/../lib/libflangFrontend.so.19.0git+0x148fef)
 #9 0x00007f98cbdcbba0 Fortran::frontend::CodeGenAction::executeAction() (/work/prbhanda/git/bhandarkar-pranav/build/install_18.0-0-25\
_Jan_24_23_23/bin/../lib/libflangFrontend.so.19.0git+0x14bba0)
#10 0x00007f98cbcdb091 Fortran::frontend::FrontendAction::execute() (/work/prbhanda/git/bhandarkar-pranav/build/install_18.0-0-25_Jan_\
24_23_23/bin/../lib/libflangFrontend.so.19.0git+0x5b091)
#11 0x00007f98cbccb2e3 Fortran::frontend::CompilerInstance::executeAction(Fortran::frontend::FrontendAction&) (/work/prbhanda/git/bhan\
darkar-pranav/build/install_18.0-0-25_Jan_24_23_23/bin/../lib/libflangFrontend.so.19.0git+0x4b2e3)
#12 0x00007f98cc5a6b1b Fortran::frontend::executeCompilerInvocation(Fortran::frontend::CompilerInstance*) (/work/prbhanda/git/bhandark\
ar-pranav/build/install_18.0-0-25_Jan_24_23_23/bin/../lib/libflangFrontendTool.so.19.0git+0x4b1b)
#13 0x000055cd42a84c38 fc1_main(llvm::ArrayRef<char const*>, char const*) (/work/prbhanda/git/bhandarkar-pranav/build/install_18.0-0-2\
5_Jan_24_23_23/bin/flang-new+0x5c38)
#14 0x000055cd42a8335d main (/work/prbhanda/git/bhandarkar-pranav/build/install_18.0-0-25_Jan_24_23_23/bin/flang-new+0x435d)
#15 0x00007f98c8c29d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#16 0x00007f98c8c29e40 call_init ./csu/../csu/libc-start.c:128:20
#17 0x00007f98c8c29e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#18 0x000055cd42a83865 _start (/work/prbhanda/git/bhandarkar-pranav/build/install_18.0-0-25_Jan_24_23_23/bin/flang-new+0x4865)
flang-new: error: unable to execute command: Segmentation fault (core dumped)
flang-new: error: flang frontend command failed due to signal (use -v to see invocation)
flang-new version 19.0.0git (git@github.com:bhandarkar-pranav/llvm-project.git 2e0ddfc1638597e0d08fb028fdd6754387732107)
Target: x86_64-unknown-linux-gnu
Thread model: posix
```

### PR DESCRIPTION
**Pull request by bhandarkar-pranav**
This patch fixes an issue in the conversion of the `omp` MLIR dialect to LLVMIR using the OpenMPIRBuilder seen during offloading.
The `finalize` method in OpenMPIRBuilder outlines regions previoulsy marked for outlining (For example, when processing `omp.task` ops). To do so, it uses a datastructure that holds `BasicBlock`s that need to be outline.
However, if these regions belong to host code in a device module then these regions will have been removed when the `omp.declare_target` attribute is converted. As a result the `finalize` method ends up accessing bad data and crashes.

Fixes https://github.com/llvm/llvm-project/issues/84606

### CODE REVIEW

**Review by skatrak — COMMENTED** [review]
Thanks for looking into this Pranav. It looks related to this downstream patch by @DominikAdamski: https://github.com/ROCm/llvm-project/pull/40.

It's yet another manifestation of the issues we have due to having to keep host code when compiling for the device just so that we're able to lower target regions inside. When we had the early outlining pass we didn't have this problem, because only device code remained by the time we lowered to LLVM IR. Now we're lowering all of the host code that exists in functions with an `omp.target` inside to then remove it later, which causes problems like what you're fixing here.

I think the long term solution for this is to actually remove host code from device compilation early (re-introducing some sort of early outlining, or at least somehow removing host operations in a pass while keeping target regions untouched). It seems difficult to selectively detect and skip lowering for these operations in advance in all cases, like Dominik's PR tries to do. And fixing the IR after lowering, like this patch does, is very difficult to get right for all situations, since lowering host operations might result in the creation of some OMPIRBuilder state used after the function has been deleted and this will change over time.

So I think this PR looks good as a short term solution to the issue, but we need to figure out a better plan long term. I'd like to hear from others' opinions about this more general issue. I don't oppose merging this PR, but I'll leave the approval to someone else.

**Comment by bhandarkar-pranav** [comment]
> Thanks for looking into this Pranav. It looks related to this downstream patch by @DominikAdamski: [ROCm#40](https://github.com/ROCm/llvm-project/pull/40).
> 
> It's yet another manifestation of the issues we have due to having to keep host code when compiling for the device just so that we're able to lower target regions inside. When we had the early outlining pass we didn't have this problem, because only device code remained by the time we lowered to LLVM IR. Now we're lowering all of the host code that exists in functions with an `omp.target` inside to then remove it later, which causes problems like what you're fixing here.
> 
> I think the long term solution for this is to actually remove host code from device compilation early (re-introducing some sort of early outlining, or at least somehow removing host operations in a pass while keeping target regions untouched). It seems difficult to selectively detect and skip lowering for these operations in advance in all cases, like Dominik's PR tries to do. And fixing the IR after lowering, like this patch does, is very difficult to get right for all situations, since lowering host operations might result in the creation of some OMPIRBuilder state used after the function has been deleted and this will change over time.
> 
> So I think this PR looks good as a short term solution to the issue, but we need to figure out a better plan long term. I'd like to hear from others' opinions about this more general issue. I don't oppose merging this PR, but I'll leave the approval to someone else.

Thanks for pointing me to https://github.com/ROCm/llvm-project/pull/40 @skatrak 
I agree this fix is very awkward and one can never be sure that all the ``OMPIRBuilder`` state has been cleaned up properly. The best is to avoid creating unnecessary state in the first place. I am curious about why early outlining was removed in the first place because on the surface early outlining makes sense to me (There must have been a good reason). In the meantime, I'll follow the discussion on @DominikAdamski's PR.

Do you suggest, I close this PR then? I need to fix this problem so I can continue working on lowering the `depend` clause. The lowering of the `depend` clause goes through a transformation (https://github.com/llvm/llvm-project/pull/83966) that converts
```
omp.target depend(...)
```
to 
```
omp.task depend(...) {
 omp.target {
 }
}
```
So, we will end up introducing host code anyway even in a function that has only the ``omp.target`` in it

**Comment by skatrak** [comment]
My preference would be to still merge it, since it does fix some problems. We just need to be aware of its limitations and start thinking of a better way to deal with the root cause.

I think @agozillon and @jsjodin are more familiar with the reasons for the early outlining to be removed, but I seem to remember that it may have been done too early, resulting in complications when dealing with map clauses. I might be misremembering, though, so they might be able to correct me on that.

File: mlir/test/Target/LLVMIR/openmp-task-target-device.mlir
Comment (DominikAdamski) [inline mlir/test/Target/LLVMIR/openmp-task-target-device.mlir:1]: Could you provide a more detailed test that proves no host-related LLVM-IR is present in the generated target LLVM IR?

**Review by DominikAdamski — COMMENTED** [review]
I think that an MLIR pass, which operates on LLVM/OpenMP dialects and outlines target regions, is a good idea. I opt for a pass for the LLVM dialect because this dialect will be common for both Fortran and future C/C++ MLIR code. Its main advantage will be the removal of redundant LLVM IR code generation.

Currently, we remove this redundant code in the `convertDeclareTargetAttr` function. We had a discussion about whether this approach is acceptable, and we decided to accept it (see https://reviews.llvm.org/D147641 ). If we remove the original host function, we should also remove the information about planned LLVM IR outlining. Having said that, I tend to accept this patch as a temporary solution.

**Comment by bhandarkar-pranav** [comment]
Closing this PR because an alternative, https://github.com/llvm/llvm-project/pull/85239, was merged.

**Pull request closed without merging**